build-tools/xa-prep-tasks/Xamarin.Android.BuildTools.PrepTasks/ProcessLogcatTiming.cs

```diff
@@ -0,0 +1,89 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Text.RegularExpressions;
+
+using Microsoft.Build.Framework;
+using Microsoft.Build.Utilities;
+
+namespace Xamarin.Android.BuildTools.PrepTasks
+{
+	public class ProcessLogcatTiming : Task
+	{
+		public string Logcat { get; set; }
+
+		public string Application { get; set; }
+
+		public string ResultsFile { get; set; }
+
+		public override bool Execute ()
+		{
+			using (var reader = new StreamReader (Logcat)) {
+				string line;
+				int pid = -1;
+				var procStartRegex = new Regex ($"^(\\d+-\\d+\\s+[\\d\\:\\.]+)\\s+.*ActivityManager: Start proc.*for added application {Application}\\: pid=(\\d+)");
+				Regex timingRegex = null;
+				var runtimeInitRegex = new Regex ("Runtime\\.init\\: end native-to-managed");
+				DateTime start = DateTime.Now;
+				DateTime last = start;
+				DateTime initEnd = start;
+
+				while ((line = reader.ReadLine ()) != null) {
+					if (pid == -1) {
+						var match = procStartRegex.Match (line);
+						if (match.Success) {
+							start = ParseTime (match.Groups [1].Value);
+							pid = Int32.Parse (match.Groups [2].Value);
+							Log.LogMessage (MessageImportance.Low, $"Time:      0ms process start, application: '{Application}' PID: {pid}");
+							timingRegex = new Regex ($"^(\\d+-\\d+\\s+[\\d\\:\\.]+)\\s+{pid}\\s+.*I monodroid-timing\\:\\s(.*)$");
+						}
+					} else {
+						var match = timingRegex.Match (line);
+						if (match.Success) {
+							var time = ParseTime (match.Groups [1].Value);
+							var span = time - start;
+							Log.LogMessage (MessageImportance.Low, $"Time: {span.TotalMilliseconds.ToString ().PadLeft (6)}ms Message: {match.Groups [2].Value}");
+
+							match = runtimeInitRegex.Match (match.Groups [2].Value);
+							if (match.Success)
+								initEnd = time;
+							last = time;
+						}
+					}
+				}
+
+				if (pid != -1) {
+					Log.LogMessage (MessageImportance.Normal, " -- Performance summary --");
+					Log.LogMessage (MessageImportance.Normal, $"Runtime init end: {(initEnd - start).TotalMilliseconds}ms");
+					Log.LogMessage (MessageImportance.Normal, $"Last timing message: {(last - start).TotalMilliseconds}ms");
+					if (ResultsFile != null) {
+						File.WriteAllText (Path.Combine (Path.GetDirectoryName (ResultsFile), $"Test-{Application}-times.csv"),
+						                   $"init,last\n{(initEnd - start).TotalMilliseconds},{(last - start).TotalMilliseconds}");
+					}
+				} else
+					Log.LogWarning ("Wasn't able to collect the performance data");
+
+				reader.Close ();
+			}
+
+			return true;
+		}
+
+		static Regex timeRegex = new Regex ("(\\d+)-(\\d+)\\s+(\\d+)\\:(\\d+)\\:(\\d+)\\.(\\d+)");
+		DateTime ParseTime (string s)
+		{
+			var match = timeRegex.Match (s);
+			if (match.Success)
```

Comment:
Again, please invert so that the "shorter" case is in the `if` block:

```csharp
if (!match.Success)
    throw new InvalidOperationException ($"Unable to parse time: '{s}'");
return new DateTime (...);
```